simpleline/render/renderer.py

```diff
@@ -0,0 +1,382 @@
+            raise KeyError("Can't run multiple input threads at the same time!")
+
+        input_queue = queue.Queue()
+        self._input_thread = threading.Thread(target=self._thread_input, name="InputThread",
+                                              args=(input_queue, prompt, hidden))
+        self._input_thread.daemon = True
+        self._input_thread.start()
+        self._event_loop.process_signals(return_after=InputReadySignal)
+        return input_queue.get()  # return the user input
+
+    def input(self, args, key):
+        """Method called internally to process unhandled input key presses.
+
+        Also handles the main quit and close commands.
+
+        :param args: optional argument passed from switch_screen calls
+        :type args: anything
+
+        :param key: the string entered by user
+        :type key: str
+
+        :return: True if key was processed, False if it was not recognized
+        :rtype: True|False
+        """
+        # delegate the handling to active screen first
+        if not self._screen_stack.is_empty():
+            try:
+                key = self._screen_stack.pop(False).ui_screen.input(args, key)
+                if key == INPUT_PROCESSED:
+                    return True
+                elif key == INPUT_DISCARDED:
+                    return False
+            except ExitMainLoop:
+                raise
+            except Exception:    # pylint: disable=broad-except
+                self._event_loop.enqueue_signal(ExceptionSignal(self))
+                return False
+
+            # global refresh command
+            if key == Prompt.REFRESH:
+                self.redraw()
+                return True
+
+            # global close command
+            if key == Prompt.CONTINUE:
+                self.close_screen()
+                return True
+
+            # global quit command
+            if key == Prompt.QUIT:
+                if self.quit_screen:
+                    quit_screen = self.quit_screen
+                    d = quit_screen(self, self._quit_message)
+                    self.switch_screen_modal(d)
+                    if d.answer:
+                        raise ExitMainLoop()
+                else:
+                    raise ExitMainLoop()
+                return True
+
+        return False
+
+    def _thread_input(self, queue_instance, prompt, hidden):
+        """This method is responsible for interruptable user input.
+
+        It is expected to be used in a thread started on demand
+        and returns the input via the communication Queue.
+
+        :param queue_instance: communication queue_instance to be used
+        :type queue_instance: queue.Queue instance
+
+        :param prompt: prompt to be displayed
+        :type prompt: Prompt instance
+
+        :param hidden: whether typed characters should be echoed or not
+        :type hidden: bool
+        """
+        if hidden:
+            data = getpass.getpass(prompt)
```

Comment:
Just a note for a future pull request - "built-in" simpleline in the Anaconda source tree recently got support for providing a custom getpass implementation (to support being able to enter passwords from an arbitrary console in Initial Setup), so we should eventually also support it here. :)